staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -2982,6 +2982,58 @@ type PodSpec struct {
 	// Optional: Defaults to true.
 	// +optional
 	EnableServiceLinks *bool `json:"enableServiceLinks,omitempty" protobuf:"varint,30,opt,name=enableServiceLinks"`
+	// TopologySpreadConstraints describes how a group of pods ought to spread across topology
+	// domains. Scheduler will schedule pods in a way which abides by the constraints.
+	// This field is alpha-level and is only honored by clusters that enables the EvenPodsSpread
+	// feature.
+	// +optional
+	// +patchMergeKey=topologyKey
+	// +patchStrategy=merge
+	TopologySpreadConstraints []TopologySpreadConstraint `json:"topologySpreadConstraints,omitempty" patchStrategy:"merge" patchMergeKey:"topologyKey" protobuf:"bytes,31,opt,name=topologySpreadConstraints"`
+}
+
+type UnsatisfiableConstraintAction string
+
+const (
+	// DoNotSchedule instructs the scheduler not to schedule the pod
+	// when constraints are not satisfied.
+	DoNotSchedule UnsatisfiableConstraintAction = "DoNotSchedule"
+	// ScheduleAnyway instructs the scheduler to schedule the pod
+	// even if constraints are not satisfied.
+	ScheduleAnyway UnsatisfiableConstraintAction = "ScheduleAnyway"
+)
+
+// TopologySpreadConstraint specifies how to spread matching pods among the given topology.
+type TopologySpreadConstraint struct {
+	// MaxSkew describes the degree to which pods may be unevenly distributed.
+	// It's the maximum permitted difference between the number of matching pods in
+	// any two topology domains of a given topology type.
+	// For example, in a 3-zone cluster, currently pods with the same labelSelector
+	// are "spread" such that zone1 and zone2 each have one pod, but not zone3.
+	// - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+	// scheduling it onto zone1(zone2) would make the ActualSkew(2) violate MaxSkew(1)
+	// - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+	// It's a required field. Default value is 1 and 0 is not allowed.
+	MaxSkew int32 `json:"maxSkew" protobuf:"varint,1,opt,name=maxSkew"`
+	// TopologyKey is the key of node labels. Nodes that have a label with this key
+	// and identical values are considered to be in the same topology.
+	// We consider each <key, value> as a "bucket", and try to put balanced number
+	// of pods into each bucket.
+	// It's a required field.
+	TopologyKey string `json:"topologyKey" protobuf:"bytes,2,opt,name=topologyKey"`
+	// WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy
+	// the spread constraint.
+	// - DoNotSchedule (default) tells the scheduler not to schedule it
+	// - ScheduleAnyway tells the scheduler to still schedule it
+	// Note: it's considered as "Unsatisfiable" only when actual skew on all nodes
+	// exceeds "MaxSkew".
+	// It's a required field.
+	WhenUnsatisfiable UnsatisfiableConstraintAction `json:"whenUnsatisfiable" protobuf:"bytes,3,opt,name=whenUnsatisfiable,casttype=UnsatisfiableConstraintAction"`
```

Comment:
Is there a policy for what to do if the pods become imbalanced via attrition?

Reply:
@lavalamp this Even Pod Spreading feature only applies to scheduling time decisions. Runtime even spreading due to any number of issues like drain, auto scaling ,etc is not covered. I am making a change in descheduler to incorporate the same policy for runtime. That means, when maxSkew is not honored for a topokey, the descheduler will evict some pods and assume the scheduling will do the right thing.

Having said, i would be interested in knowing if container orchestrators like . Borg, Tupperware handle runtime spreading ? Would that be something we could incoporate into k/k ?

Reply:
If a descheduler is expected to take action on this as well, the comments should document that behavior. (Although that seems a little weird to me, I'd expect this to only be relevant if we are scaling down a deployment. If e.g. 2 pods in one zone fail, I'd expect users to be quite angry if the descheduler then killed 1 pod in each of their other zones!!)

Reply:
> Is there a policy for what to do if the pods become imbalanced via attrition?

Default scheduler does nothing on it - like what we are doing all the time.

However, @lavalamp you're right, the only case I do want to achieve is the scale-down case - arbitrary taking out a replica is not an ideal case - I'm going to tackle it down in Beta either in scheduler or workload controller side.

Reply:
@lavalamp  failed pods will not be accounted for in descheduler. Descheduler will only look at evictable pods , which means pods that are running and ready and are not daemonset pods. 

We probably need to document this better, but even pod spreading is only applied when a new pod is scheduled, it should be as far apart in the domain as possible, to other already running pods

Descheduler policy, will look at already running pods, divide them into buckets and see if the spread doesnt follow maxSkew, start evicting some pods from more full buckets and assume scheduling will fix them

@bgrant0607 may be has insight on the Borg question, Does it handle runtime spreading across domains ?

Reply:
Please answer the question in the comments :)

Reply:
@lavalamp probably i misunderstood your question. I think attrition can only happen by scaling down. The scaling down case can be solved in two ways

1: fix the scaling down in workload controller to account for topo constraints. I think this is less elegant and spreads the constraint knowledge in both scheduler and workload controller
2: let the descheduler handle the eviction and the scheduler will automagically fix this. although this does require dependency on an external component

>  If e.g. 2 pods in one zone fail, I'd expect users to be quite angry if the descheduler then killed 1 pod in each of their other zones!!)

What do you mean by 2 pods in one zone fail ? Do you mean 2 pods are pending ?